File: handy-java/quiz/q19_move_zeroes_easy.java
import java.util.Arrays;

/**
 * Q19) Move Zeroes
 *
 * AI-BEST:
 * Two pointers (read/write) in-place.
 * Time: O(n), Space: O(1)
 */
public class q19_move_zeroes_easy {
    static int[] moveZeroesBest(int[] nums) {
        // Step 1: Write non-zero values to front in original order.
        // Why: this preserves stable ordering of non-zero elements.
        int write = 0;
        for (int read = 0; read < nums.length; read++) {
            if (nums[read] != 0) {
                nums[write] = nums[read];
                write++;
            }
        }

        // Step 2: Fill the remaining tail with zeroes.
        // Why: everything beyond write is logically empty after compaction.
        while (write < nums.length) {
            nums[write] = 0;
            write++;
        }

        // Step 3: Return for demo display.
        return nums;
    }

    public static void main(String[] args) {
        System.out.println("Q19: Move Zeroes");
        int[][] cases = {
                {0, 1, 0, 3, 12},
                {0},
                {4, 1, 2},
                {0, 0, 0, 5}
        };

        for (int[] nums : cases) {
            int[] copy = Arrays.copyOf(nums, nums.length);
            System.out.println("Input: nums=" + Arrays.toString(nums));
            System.out.println("  BEST: " + Arrays.toString(moveZeroesBest(copy)));
        }
    }
}

File: handy-ts/quiz/q19_move_zeroes_easy.ts
/**
 * Q19) Move Zeroes
 *
 * AI-BEST:
 * Stable in-place two-pointer compaction.
 * Time: O(n), Space: O(1)
 */

function moveZeroesBest(nums: number[]): number[] {
  let write = 0;

  // Step 1: Keep a write pointer because we want the next non-zero value to land in the next open slot.
  for (let read = 0; read < nums.length; read++) {
    if (nums[read] !== 0) {
      // Step 2: Copy each non-zero forward because preserving order matters for this problem.
      nums[write] = nums[read];
      write++;
    }
  }

  // Step 3: Fill the remainder with zeroes because every non-zero has already been compacted to the front.
  while (write < nums.length) {
    nums[write] = 0;
    write++;
  }

  return nums;
}

function runDemo(): void {
  console.log("Q19: Move Zeroes");
  const cases = [
    [0, 1, 0, 3, 12],
    [0, 0, 1],
    [1, 2, 3],
  ];

  for (const nums of cases) {
    const working = [...nums];
    console.log(`Input: nums=${JSON.stringify(nums)}`);
    console.log("  BEST:", moveZeroesBest(working));
  }
}

runDemo();

export {};
